### ISSUE TITLE
make gettext  error

### ISSUE THREAD

**Issue by meshiguge** [sphinx-doc/sphinx#3510]
### Problem
- <hi, when use  `make gettext `  in my docs, "startswith"  error caused: 

> Exception occurred:
>   File "/Users/xx/anaconda/lib/python2.7/posixpath.py", line 54, in isabs
>     return s.startswith('/')
> AttributeError: StringList instance has no attribute 'startswith'
> The full traceback has been saved in /var/folders/hq/x053v2n52fbbkyd99d_46ywh0000gn/T/sphinx-err-wgSdEl.log, if you want to report the issue to the developers.
> Please also report this if it was a user error, so that a better error message can be provided next time.
> A bug report can be filed in the tracker at <https://bitbucket.org/birkenfeld/sphinx/issues/>. Thanks!
> make: *** [gettext] Error 1

anyone knows how to do ? thanks 
>

#### Procedure to reproduce the problem
```
<make gettext>
```

#### Error logs / results
```
<ackages/sphinx/ext/oldcmarkup.pyc
Traceback (most recent call last):
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/cmdline.py", line 254, in main
app.build(force_all, filenames)
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/application.py", line 220, in build
self.builder.build_update()
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/builders/init.py", line 214, in build_update
'out of date' % len(to_build))
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 177, in build
I18nBuilder.build(self, docnames, summary, method)
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/builders/init.py", line 279, in build
self.finish()
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 206, in finish
for source, line, _ in positions))
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 206, in 
for source, line, _ in positions))
File "/Users/xx/anaconda/lib/python2.7/site-packages/sphinx/util/osutil.py", line 160, in safe_relpath
return relpath(path, start)
File "/Users/xx/anaconda/lib/python2.7/posixpath.py", line 431, in relpath
path_list = [x for x in abspath(path).split(sep) if x]
File "/Users/xx/anaconda/lib/python2.7/posixpath.py", line 360, in abspath
if not isabs(path):
File "/Users/xx/anaconda/lib/python2.7/posixpath.py", line 54, in isabs
return s.startswith('/')
AttributeError: StringList instance has no attribute 'startswith'>
```
- <public link of unexpected result if you have>

### Environment info
- OS: <mac >
- Python version: anaconda/lib/python2.7
- Sphinx version: Sphinx v1.2.3

**Comment by tk0miya** [comment]
>Sphinx version: Sphinx v1.2.3

Could you update your Sphinx to latest one (v1.5.3) and try it again?
v1.2.3 is too long and no longer maintained.

**Comment by meshiguge** [comment]
@tk0miya, KOMIYA  san, I have updated to latest one, and error still caused.

**Comment by tk0miya** [comment]
Could you paste the stacktrace of latest Sphinx?
And could you share your project if possible? I'd like to reproduce it in my local.

**Comment by meshiguge** [comment]
@tk0miya actually, I want  to "make gettext" for the `docs` in [Lasagne](https://github.com/Lasagne/Lasagne) , and exception always caused.

**Comment by meshiguge** [comment]
Exception occurred:
  File "/Users/xx/anaconda/lib/python2.7/posixpath.py", line 54, in isabs
    return s.startswith('/')
AttributeError: StringList instance has no attribute 'startswith'
The full traceback has been saved in /var/folders/hq/x053v2n52fbbkyd99d_46ywh0000gn/T/sphinx-err-qs42mB.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [gettext] Error 1

**Comment by f0k** [comment]
Can confirm that this happens in Lasagne, both with Sphinx 1.2.3 and 1.5.3. 
To reproduce (hope this covers everything):
```bash
cd /tmp
virtualenv lasagne
cd lasagne
. lasagne/activate
git clone https://github.com/Lasagne/Lasagne.git
cd Lasagne
pip install -r requirements-dev.txt
pip install Sphinx==1.5.3
cd docs
make gettext
```

<details>
<summary>stack trace</summary>

```
# Sphinx version: 1.5.3
# Python version: 2.7.6 (CPython)
# Docutils version: 0.13.1 release
# Jinja2 version: 2.9.5
# Last messages:
#   writing output... [ 83%] modules/utils
#   writing output... [ 87%] user/custom_layers
#   writing output... [ 90%] user/development
#   writing output... [ 93%] user/installation
#   writing output... [ 96%] user/layers
#   writing output... [100%] user/tutorial
#   
#   <autosummary>:1: WARNING: citation not found: R30
#   writing message catalogs... [ 33%] index
#   writing message catalogs... [ 66%] modules
# Loaded extensions:
#   sphinx.ext.mathjax (1.5.3) from /run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/ext/mathjax.pyc
#   sphinx.ext.autodoc (1.5.3) from /run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/ext/autodoc.pyc
#   sphinx.ext.doctest (1.5.3) from /run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/ext/doctest.pyc
#   alabaster (0.7.10) from /run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/alabaster/__init__.pyc
#   sphinx.ext.linkcode (1.5.3) from /run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/ext/linkcode.pyc
#   numpydoc (unknown version) from /usr/lib/python2.7/dist-packages/numpydoc/__init__.pyc
#   sphinx.ext.autosummary (1.5.3) from /run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/ext/autosummary/__init__.pyc
Traceback (most recent call last):
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/cmdline.py", line 296, in main
    app.build(opts.force_all, filenames)
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/application.py", line 333, in build
    self.builder.build_update()
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 251, in build_update
    'out of date' % len(to_build))
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 208, in build
    I18nBuilder.build(self, docnames, summary, method)
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 325, in build
    self.finish()
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 238, in finish
    for source, line, _ in positions))
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/builders/gettext.py", line 238, in <genexpr>
    for source, line, _ in positions))
  File "/run/shm/sphinx/sphinx/local/lib/python2.7/site-packages/sphinx/util/osutil.py", line 192, in safe_relpath
    return os.path.relpath(path, start)
  File "/run/shm/sphinx/sphinx/lib/python2.7/posixpath.py", line 438, in relpath
    path_list = [x for x in abspath(path).split(sep) if x]
  File "/run/shm/sphinx/sphinx/lib/python2.7/posixpath.py", line 367, in abspath
    if not isabs(path):
  File "/run/shm/sphinx/sphinx/lib/python2.7/posixpath.py", line 61, in isabs
    return s.startswith('/')
AttributeError: StringList instance has no attribute 'startswith'
```
</details>

**Comment by tk0miya** [comment]
As a quick look, it seems this is a bug of numpydoc.
I can reproduce bug with following project:
```
# conf.py
master_doc = 'index'
extensions = ['sphinx.ext.autodoc', 'numpydoc']
```
```
index.rst
=========

.. autofunction:: sphinx.main()

   hello world
```
```
$ sphinx-build -T -b gettext -d _build/doctree . _build/html
Running Sphinx v1.6
loading pickled environment... not yet created
building [gettext]: targets for 0 template files
building [gettext]: targets for 1 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index
writing message catalogs... [100%] index
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/cmdline.py", line 306, in main
    app.build(opts.force_all, filenames)
  File "/Users/tkomiya/work/sphinx/sphinx/application.py", line 361, in build
    self.builder.build_update()
  File "/Users/tkomiya/work/sphinx/sphinx/builders/__init__.py", line 279, in build_update
    'out of date' % len(to_build))
  File "/Users/tkomiya/work/sphinx/sphinx/builders/gettext.py", line 252, in build
    I18nBuilder.build(self, docnames, summary, method)
  File "/Users/tkomiya/work/sphinx/sphinx/builders/__init__.py", line 349, in build
    self.finish()
  File "/Users/tkomiya/work/sphinx/sphinx/builders/gettext.py", line 284, in finish
    for source, line, _ in positions))
  File "/Users/tkomiya/work/sphinx/sphinx/builders/gettext.py", line 284, in <genexpr>
    for source, line, _ in positions))
  File "/Users/tkomiya/work/sphinx/sphinx/util/osutil.py", line 210, in safe_relpath
    return os.path.relpath(path, start)
  File "/Users/tkomiya/work/tmp/lib/python2.7/posixpath.py", line 431, in relpath
    path_list = [x for x in abspath(path).split(sep) if x]
  File "/Users/tkomiya/work/tmp/lib/python2.7/posixpath.py", line 360, in abspath
    if not isabs(path):
  File "/Users/tkomiya/work/tmp/lib/python2.7/posixpath.py", line 54, in isabs
    return s.startswith('/')
AttributeError: StringList instance has no attribute 'startswith'

Exception occurred:
  File "/Users/tkomiya/work/tmp/lib/python2.7/posixpath.py", line 54, in isabs
    return s.startswith('/')
AttributeError: StringList instance has no attribute 'startswith'
The full traceback has been saved in /var/folders/rj/80_5mv452nsfnhkvkyqjwj040000gp/T/sphinx-err-iNmNCf.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

It works fine if I remove `numpydoc` extension.
Please report this to the numpydoc project.

Thanks,

**Comment by tk0miya** [comment]
@meshiguge Did you use `numpydoc` extension? If true, it might be a bug of the extension.

**Comment by f0k** [comment]
Indeed, removing `numpydoc` from Lasagne's Sphinx configuration gets rid of the exception. Thank you for pointing us in the right direction! Updating numpydoc to the latest version doesn't help. Seems they do something that plays well with regular documentation building in Sphinx, but not with gettext.

Now the question is if the error is in numpydoc or in the gettext builder. With `make SPHINXOPTS=-P gettext`, I can see that the error occurs here:
https://github.com/sphinx-doc/sphinx/blob/1.5.3/sphinx/builders/gettext.py#L237
`source` is expected to be a string, but here it is a `docutils.state_machine.StringList`, a list of the lines of a particular docstring and the corresponding file and line it was taken from. Is this a hack used by numpydoc, or something that gettext should support?

FWIW, the same bug had been reported earlier in https://github.com/sphinx-doc/sphinx/issues/915, found to be related to numpydoc, and not followed further.

**Comment by tk0miya** [comment]
I still don't understand how gettext builder work well, but I feel it's a bug of gettext.
`StringList` is usually used only for internal use (mainly on directive processing). So it's strange to see it outside of directive.

**Comment by tk0miya** [comment]
I'd investigated carefully, but I can't reproduce without numpydoc. So I think this might be a bug of numpydoc.
So I close this now.
Thanks,

**Comment by f0k** [comment]
For anybody searching for this error, this has been fixed with numpydoc 0.7 (released June 20th, 2017).
